client/lib/track-element-size/index.ts

```diff
@@ -0,0 +1,118 @@
+/**
+ * External dependencies
+ */
+import { useRef, useEffect, useState, useCallback } from 'react';
+import { throttle } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import afterLayoutFlush from 'lib/after-layout-flush';
+
+const THROTTLE_RATE = 200;
+
+// Indexable version of DOMRect / ClientRect.
+interface UseWindowResizeRect {
+	[index: string]: number | undefined;
+	top: number;
+	right: number;
+	bottom: number;
+	left: number;
+	width: number;
+	height: number;
+	x?: number;
+	y?: number;
+}
+
+// Function type for the expected callback for `useWindowResizeCallback`.
+interface UseWindowResizeCallback {
+	( node: Element | null | undefined ): void;
+}
+
+/**
+ * React hook that subscribes a consumer to changes to the bounding client rect of an element, based
+ * on window resize events.
+ * Does not notify when an element resizes due to reasons other than the window resizing.
+ * Uses throttling on the events, to avoid making changes too often.
+ *
+ * @param callback The function to call back on changes. Takes a single parameter: `boundingClientRect`.
+ *
+ * @returns The ref to be set on the consumer component.
+ */
+export function useWindowResizeCallback(
+	callback: ( boundingClientRect: UseWindowResizeRect ) => any
+) {
+	const lastRect = useRef< UseWindowResizeRect | null >( null );
+	const [ element, setElement ] = useState< Element | null >( null );
+
+	const callbackRef: UseWindowResizeCallback = useCallback( node => {
+		if ( node ) {
+			setElement( node );
+		}
+	}, [] );
+
+	useEffect( () => {
+		// Notify consumer of bounding client rect change.
+		const handleRectChange = ( rect: DOMRect | ClientRect ) => {
+			if ( rect ) {
+				const jsonRect = 'toJSON' in rect ? rect.toJSON() : rect;
```

Comment:
I was rather surprised to see that `rect.toJSON()` has a return type `any`. I had trouble finding any documentation on it, do you have any links?

Reply:
After some considered thought, I think I'll drop the JSON conversion altogether. It's not that idiomatic in JS, where duck-typing is king.